pinot-spi/src/test/java/org/apache/pinot/spi/plugin/ClassLoaderTest.java

```diff
@@ -0,0 +1,226 @@
+    Assert.assertNotNull(pluginManager.createInstance(pluginName, _yammerMetricsRegistry));
+
+    Assert.assertNotNull(pluginManager.loadClass(pluginName, _spiTimeUtils),
+        "Expected o.a.p.spi.utils.TimeUtils to be available via yammer-realm");
+    Assert.assertEquals(pluginManager.loadClass(pluginName, _spiTimeUtils)
+            .getProtectionDomain().getCodeSource().getLocation(),
+        Path.of("target/classes").toUri().toURL(),
+        "Expected o.a.p.spi.utils.TimeUtils to be loaded from pinot-realm");
+
+    Assert.assertNotNull(pluginManager.loadClass(pluginName, _commonsMathUtils),
+        "o.a.c.m.u.MathUtils is dependency of pinot-spi, must be accessible for unlimited plugins");
+
+    Assert.assertTrue(pluginManager.loadClass(pluginName, _commonsIOUtils).getProtectionDomain()
+            .getCodeSource().getLocation().getPath().endsWith("pinot-yammer-0.10.0-shaded.jar"),
+        "This is self-first, so class should come from pinot-yammer realm");
+
+    Assert.assertThrows("Class is part of a different plugin, so should not be accessible",
+        ClassNotFoundException.class, () -> pluginManager.loadClass(pluginName, _dropwizardMetricsRegistry));
+  }
+
+  @Test
+  public void assemblyBasedRealm() throws Exception {
+    final String pluginName = "assemblybased-pinot-plugin";
+
+    // Create fresh pluginManager, so it can pick up the system property for the pluginDirectory
+    PluginManager pluginManager = new PluginManager();
+
+    // pinot-yammer is an unlimited plugin, meaning it has access to every class from the pinot realm
+    Assert.assertTrue(
+        Files.exists(_pluginsDirectory.resolve(
+            pluginName + "/classes/org/apache/pinot/plugin/metrics/yammer/YammerMetricsRegistry.class")),
+        "Class not found. Run 'mvn -f pinot-spi/pom.xml generate-test-resources' once to prepare this artifact");
+    Assert.assertNotNull(pluginManager.loadClass(pluginName, _yammerMetricsRegistry));
+    Assert.assertNotNull(pluginManager.createInstance(pluginName, _yammerMetricsRegistry));
+
+    Assert.assertNotNull(pluginManager.loadClass(pluginName, _spiTimeUtils),
+        "Expected o.a.p.spi.utils.TimeUtils to be available via yammer-realm");
+    Assert.assertEquals(pluginManager.loadClass("pinot-dropwizard", _spiTimeUtils)
+            .getProtectionDomain().getCodeSource().getLocation(),
+        Path.of("target/classes").toUri().toURL(),
+        "Expected o.a.p.spi.utils.TimeUtils to be loaded from pinot-realm");
+
+    Assert.assertThrows("Class is dependency of pinot-spi, but is not an exported package",
+        ClassNotFoundException.class, () -> pluginManager.loadClass(
+            "pinot-dropwizard", _commonsMathUtils));
+
+    Assert.assertTrue(pluginManager.loadClass(pluginName, _commonsIOUtils).getProtectionDomain()
+            .getCodeSource().getLocation().getPath().endsWith("commons-io-2.11.0.jar"),
```

Comment:
This is causing #14019 to fail. Is there a way to avoid hard-code the version?

Reply:
there are more jars here with a hardcoded version, just to ensure that if a project dependency is updated, this test won't fail. In #14019 the test-resource is updated (so not a dependency), so it makes sense that is fails. 
What if one could avoid the hardcoded version and the upgrade would make the test fail? That will be very hard to analyse.

Reply:
I'm with Robert here. It looks like a problem on dependanbot, which should ignore this `artifactItem`

Reply:
You are correct. I didn't notice dependabot is trying to upgrade the hard-coded version in test release settings